### ISSUE TITLE
limited gpg path length may clash with mkosi.workspace path length

### ISSUE THREAD

**Issue by nopeslide** [systemd/mkosi#749]
gpg seems to construct a socket by using `struct sockaddr_un` from `sys/socket.h`.
from `man unix 7`:
```
         A UNIX domain socket address is represented in the following structure:

           struct sockaddr_un {
               sa_family_t sun_family;               /* AF_UNIX */
               char        sun_path[108];            /* Pathname */
           };
```
When building an image that uses gpg while installing the distro (i.e. archlinux)
the `mkosi.workspace` path may be too long to fit into `sockaddr_un.sun_path`.
gpg fails to create the socket and the distro installation fails (without any information regarding the actual problem).
Setting a shorter `mkosi.workspace` path helps, but only mitigates the problem.

A general mitigation for this could be to use a temporary symlink in `/tmp` as root pointing to the root inside `mkosi.workspace`.

**Comment by poettering** [comment]
hmm, why does gpg need to bind a socket there? what's going on?

**Comment by nopeslide** [comment]
> hmm, why does gpg need to bind a socket there? what's going on?

I assume gpg-agent is started in server mode and communicates via Unix socket.

**Comment by poettering** [comment]
by what though? and why would it put a socket in the workspace dir?

**Comment by nopeslide** [comment]
`pacman-key` uses gpg to set up keys and trusts.
I don't know why it seems to create a socket in workspace.
I didn't check if this also fails for gpg uses by other distros.
here is the output of failing build
```
$ sudo mkosi
‣ Warning: 
Permissions of 'mkosi.rootpw' of '0644' are too open.
When creating root password files use an access mode that restricts access to the owner only.                
‣ Detaching namespace...
‣ Setting up temporary workspace.
‣ Temporary workspace in /tmp/_____________________________________________________________________________/mkosi.workspace/f9i_uzi6 is now set up.
‣ Running first (development) stage...
‣ Creating partition table...
Checking that no-one is using this disk right now ... OK

Disk /tmp/_____________________________________________________________________________/mkosi.output/.mkosi-44cibz3_: 3 GiB, 3223322624 bytes, 6295552 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes

>>> Script header accepted.
>>> Created a new GPT disklabel (GUID: BA2D1820-A315-9848-B413-B094B622FF4B).
/tmp/_____________________________________________________________________________/mkosi.output/.mkosi-44cibz3_1: Created a new partition 1 of type 'Linux root (x86-64)' and of size 3 GiB.
/tmp/_____________________________________________________________________________/mkosi.output/.mkosi-44cibz3_2: Done.

New situation:
Disklabel type: gpt
Disk identifier: BA2D1820-A315-9848-B413-B094B622FF4B

Device                                                                                                           Start     End Sectors Size Type
/tmp/_____________________________________________________________________________/mkosi.output/.mkosi-44cibz3_1  2048 6295518 6293471   3G Linux root (x86-64)

The partition table has been altered.
Syncing disks.
‣ Created partition table as /tmp/_____________________________________________________________________________/mkosi.output/.mkosi-44cibz3_.
‣ Attaching image file...
‣ Attached image file as /dev/loop0.
‣ Formatting root partition...
mke2fs 1.46.2 (28-Feb-2021)
Creating filesystem with 786432 4k blocks and 196992 inodes
Filesystem UUID: f3fb2d0a-5231-4682-ab3f-0b5467f6cff0
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376, 294912

Allocating group tables: done                            
Writing inode tables: done                            
Creating journal (16384 blocks): done
Writing superblocks and filesystem accounting information: done 

‣ Mounting image...
‣ Setting up basic OS tree...
‣ Copying in skeleton file trees...
‣ Mounting Package Cache...
‣ Installing Arch Linux...
Distribution release specification is not supported for Arch Linux, ignoring.
‣ Mounting API VFS...
gpg: /tmp/_____________________________________________________________________________/mkosi.workspace/f9i_uzi6/root/etc/pacman.d/gnupg/trustdb.gpg: trustdb created
gpg: no ultimately trusted keys found
gpg: starting migration from earlier GnuPG versions
gpg: can't connect to the agent: File name too long
gpg: error: GnuPG agent unusable. Please check that a GnuPG agent can be started.
gpg: migration aborted
==> Generating pacman master key. This may take some time.
gpg: starting migration from earlier GnuPG versions
gpg: can't connect to the agent: File name too long
gpg: error: GnuPG agent unusable. Please check that a GnuPG agent can be started.
gpg: migration aborted
gpg: Generating pacman keyring master key...
gpg: can't connect to the agent: File name too long
gpg: agent_genkey failed: No agent running
gpg: key generation failed: No agent running
gpg: Done
==> Updating trust database...
gpg: no need for a trustdb check
gpg: starting migration from earlier GnuPG versions
gpg: can't connect to the agent: File name too long
gpg: error: GnuPG agent unusable. Please check that a GnuPG agent can be started.
gpg: migration aborted
==> ERROR: There is no secret key available to sign with.
==> Use 'pacman-key --init' to generate a default secret key.
‣ Unmounting API VFS...
‣ Unmounting Package Cache...
‣ Unmounting image...
‣ Detaching image file...
Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/user/.local/lib/python3.9/site-packages/mkosi/__main__.py", line 31, in <module>
    main()
  File "/home/user/.local/lib/python3.9/site-packages/mkosi/__main__.py", line 25, in main
    run_verb(a)
  File "/home/user/.local/lib/python3.9/site-packages/mkosi/__init__.py", line 7295, in run_verb
    build_stuff(args)
  File "/home/user/.local/lib/python3.9/site-packages/mkosi/__init__.py", line 6729, in build_stuff
    raw, tar, root_hash, sshkey, split_root, split_verity, split_kernel = build_image(
  File "/home/user/.local/lib/python3.9/site-packages/mkosi/__init__.py", line 6485, in build_image
    install_distribution(args, root, do_run_build_script, cached_tree)
  File "/home/user/.local/lib/python3.9/site-packages/mkosi/__init__.py", line 3155, in install_distribution
    install[args.distribution](args, root, do_run_build_script)
  File "/usr/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/user/.local/lib/python3.9/site-packages/mkosi/__init__.py", line 3015, in install_arch
    run_pacman(root, pacman_conf, packages)
  File "/home/user/.local/lib/python3.9/site-packages/mkosi/__init__.py", line 2700, in run_pacman
    run(["pacman-key", "--config", pacman_conf, "--populate"])
  File "/home/user/.local/lib/python3.9/site-packages/mkosi/__init__.py", line 260, in run
    return subprocess.run(cmdline, check=check, stdout=stdout, stderr=stderr, **kwargs)
  File "/usr/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['pacman-key', '--config', '/tmp/_____________________________________________________________________________/mkosi.workspace/f9i_uzi6/pacman.conf', '--populate']' returned non-zero exit status 1.
```

**Comment by dvdhrm** [comment]
Just in case someone else runs into this: `pacman-key` uses `gpg` to manage its key database. All `gpg` commands require an agent, see the man-page:

```
       --use-agent
       --no-use-agent
              This is dummy option. gpg always requires the agent.
```

With `mkosi` you easily run into the 108-character limit of unix-domain-sockets. I assume other distributions are affected as well, but might employ chroots/namespaces/symlinks to reduce the exposure.

I don't see a nice fix to this, other than the one already suggested (symlink in /tmp).

**Comment by dvdhrm** [comment]
> This was fixed upstream I think with [systemd/systemd#23339](https://github.com/systemd/systemd/pull/23339) so let's close this

This is not fixed. The comment you mention fixed systemd to be able to open UDS with overlong paths, but it does not fix any external program. This is still an issue for GPG, for instance.

I am fine closing this bug, if you feel that it is not the job of mkosi to track these issues. However, note that ArchLinux as backend is currently broken in mkosi if built in a directory that has a path-length of >~30, because the internal mkosi paths would then exceed the 108-limit and break gpg-invocations in the archlinux setup.

**Comment by DaanDeMeyer** [comment]
This seems like something that should be fixed in gpg, so let's close the issue here
